src/workspace/resolver.rs

```diff
@@ -0,0 +1,373 @@
+  ///
+  /// This should only be set when a node_modules folder doesn't exist
+  /// as node resolution should be used otherwise.
+  pub pkg_json_dep_resolution: bool,
+  pub specified_import_map: Option<SpecifiedImportMap>,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpecifiedImportMap {
+  pub base_url: Url,
+  pub value: serde_json::Value,
+}
+
+#[derive(Debug, Clone)]
+pub enum MappedResolution<'a> {
+  Normal(Url),
+  ImportMap(Url),
+  PackageJson {
+    pkg_json: &'a PackageJsonRc,
+    alias: &'a str,
+    req_ref: NpmPackageReqReference,
+  },
+}
+
+impl<'a> MappedResolution<'a> {
+  pub fn into_url(self) -> Result<Url, url::ParseError> {
+    match self {
+      Self::Normal(url) | Self::ImportMap(url) => Ok(url),
+      Self::PackageJson { req_ref, .. } => Url::parse(&req_ref.to_string()),
+    }
+  }
+}
+
+#[derive(Debug, Error)]
+pub enum MappedResolutionError {
+  #[error(transparent)]
+  Specifier(#[from] SpecifierError),
+  #[error(transparent)]
+  ImportMap(#[from] ImportMapError),
+  #[error(transparent)]
+  PkgJsonDep(#[from] PackageJsonDepValueParseError),
+}
+
+impl MappedResolutionError {
+  pub fn is_unmapped_bare_specifier(&self) -> bool {
+    match self {
+      MappedResolutionError::Specifier(err) => match err {
+        SpecifierError::InvalidUrl(_) => false,
+        SpecifierError::ImportPrefixMissing { .. } => true,
+      },
+      MappedResolutionError::ImportMap(err) => match err {
+        ImportMapError::UnmappedBareSpecifier(_, _) => true,
+        ImportMapError::Other(_) => false,
+      },
+      MappedResolutionError::PkgJsonDep(_) => false,
+    }
+  }
+}
+
+#[derive(Debug)]
+pub struct WorkspaceResolver {
+  maybe_import_map: Option<ImportMapWithDiagnostics>,
+  pkg_jsons: BTreeMap<UrlRc, PkgJsonResolverFolderConfig>,
+}
+
+impl WorkspaceResolver {
+  pub(crate) async fn from_workspace<
+    TReturn: Future<Output = Result<String, AnyError>>,
+  >(
+    workspace: &Workspace,
+    options: CreateResolverOptions,
+    fetch_text: impl Fn(&Url) -> TReturn,
+  ) -> Result<Self, WorkspaceResolverCreateError> {
+    async fn resolve_import_map<
+      TReturn: Future<Output = Result<String, AnyError>>,
+    >(
+      workspace: &Workspace,
+      specified_import_map: Option<SpecifiedImportMap>,
+      fetch_text: impl Fn(&Url) -> TReturn,
+    ) -> Result<Option<ImportMapWithDiagnostics>, WorkspaceResolverCreateError>
+    {
+      let root_folder = workspace.root_folder().1;
+      let deno_jsons = workspace
+        .config_folders()
+        .iter()
+        .filter_map(|(_, f)| f.deno_json.as_ref())
+        .collect::<Vec<_>>();
+
+      let (import_map_url, import_map) = match specified_import_map {
+        // todo(THIS PR): is it ok that if someone does --import-map= that it just
+        // overwrites the entire workspace import map?
+        Some(SpecifiedImportMap {
+          base_url,
+          value: import_map,
+        }) => (base_url, import_map),
+        None => {
```

Comment:
Reminder to address this before landing. IMO - using workspace and `--import-map` should error out for the time being unless we get a valid use case. Import map should be defined in the `deno.json(c)` file for workspaces